Docs/changes to advanced tutorial (#2465)
dlt-hub/dlt: docs/website/docs/plus/getting-started/advanced_tutorial.md

---
title: Packaging a dlt+ Project tutorial
description: Using the dlt+ cli commands to package a dlt+ Project and enable secure access to data
keywords: [command line interface, cli, dlt init, dlt+, project]
---

import Link from '../../_plus_admonition.md';

<Link/>

> Packaging a dlt+ project simplifies distribution across teams or stakeholders, such as data analysts or data science teams, without requiring direct access to the project’s internal code. Once installed, the project can be used to run pipelines and access production data through a standardized Python interface.

Reviewer: ```suggestion
Packaging a dlt+ Project simplifies distribution across teams or stakeholders, such as data analysts or data science teams, without requiring direct access to the project’s internal code. Once installed, the package can be used to run pipelines and access production data through a standardized Python interface.
```

> In this tutorial, you will learn how to package your dlt+ project for reuse and distribution within your organization.

Reviewer: ```suggestion
In this tutorial, you will learn how to package your dlt+ project for reuse and distribution and make it pip-installable.
```

## Prerequisites

Before you begin, ensure the following requirements are met:

- dlt+ is installed and set up according to the [installation guide](./installation.md)
- You are familiar with the [core concepts of dlt](../../reference/explainers/how-dlt-works.md)
- You have completed the [basic project tutorial](./tutorial.md)

Additionally, install the required Python packages:

```sh
pip install pandas numpy pyarrow streamlit dlt[duckdb] uv
```

## Packaging a project

`dlt+` provides tools to help you package a project for distribution. This makes your project installable via `pip` and easier to share across your organization.

To create the project structure required for a package, add the `--package` option when initializing:

```sh
dlt project init arrow duckdb --package my_dlt_project
```

> This creates the same basic project as in the tutorial, but places it inside a module named `my_dlt_project`, and includes a basic `pyproject.toml` file following PEP standards. 

Reviewer: ```suggestion
This creates the same basic project as in the [basic tutorial](./tutorial.md), but places it inside a module named `my_dlt_project`, and includes a basic `pyproject.toml` file following PEP standards. 
```

The key difference is the module structure and the presence of the `pyproject.toml` file. 
The file includes a special entry point setting to let dlt+ discover your project:

```toml
[project.entry-points.dlt_package]
dlt-project = "my_project"
```

You can still run the pipeline as usual with the cli commands from the root folder:

```sh
dlt pipeline my_pipeline run
```

If you open the `__init__.py` file inside your project module, you'll see the full interface that users of your package will interact with.
This interface is very similar to the `current` interface used in flat (non-packaged) projects. The main difference is that it automatically uses the `access` profile by default.
You can customize the `__init__.py` file to your project's needs.
Learn more about the `current` interface and how it enables interaction with dlt+ projects [here](../features/projects.md#python_api_to_interact_with_dlt_project).

### Using the packaged project

To demonstrate how your packaged project can be used, let's simulate a real-world scenario where a data scientist installs and runs your project in a separate Python environment.

> In this example, we'll use the **uv** package manager, but the same steps apply when using **poetry** or **pip**.

Reviewer: Please, add link to the uv package manager: https://github.com/astral-sh/uv
and how to install it: https://github.com/astral-sh/uv?tab=readme-ov-file#installation

Assume your packaged dlt+ project is located at: `/Volumes/my_drive/my_folder/pyproject.toml`. 
Navigate to a new directory and initialize your project:

```sh
uv init
```

Install your packaged project directly from the local path:

```sh
uv pip install /Volumes/my_drive/my_folder
```

Your dlt+ project is now available for use in this environment.

As an exemple, create a new Python file named `test_project.py` and use your packaged project like this:

```py
# import the packaged project
import my_dlt_project

> from my_dlt_project import *

Reviewer: This import should not be here. Generally, `from something import *` is not a very good pattern, so we don't recommend it in our docs.

if __name__ == "__main__":
    # should print "access" as defined in your dlt package
    print(my_dlt_project.config().current_profile)
    # Run the pipeline from the packaged project

>     my_dlt_project.runner().run_pipeline("my_pipeline")

Reviewer: Hey, same comment here

Assume your packaged dlt+ project is located at: `/Volumes/my_drive/my_folder/pyproject.toml`. 
Navigate to a new directory and initialize your project:

```sh
uv init
```

Install your packaged project directly from the local path:

```sh
uv pip install /Volumes/my_drive/my_folder
```

Your dlt+ project is now available for use in this environment.

As an exemple, create a new Python file named `test_project.py` and use your packaged project like this:

```py
# import the packaged project
import my_dlt_project

if __name__ == "__main__":
    # should print "access" as defined in your dlt package
    print(my_dlt_project.config().current_profile)
    # Run the pipeline from the packaged project
    my_dlt_project.runner().run_pipeline("my_pipeline")
    # should list the defined destinations  
    print(my_dlt_project.config().destinations)
```
Define the environment variables it needs. Add the following lines before running your pipeline:

> ```py

Reviewer: I meant we should merge this Python script with the previous one at line 95 ;)

<Link/>

In this tutorial, you will explore advanced features of a dlt+ project. By the end, you will be able to:
* Understand how dlt+ automatically creates implicit entities when not explicitly defined—and how to prevent this behavior
* Use dlt+ datasets to control access to specific datasets and destinations 
* Interact with the dlt+ project context and reference entities directly from your code
* Package a dlt+ project for distribution to other teams or stakeholders within your organization

## Prerequisites

Before you begin, ensure the following requirements are met:

- dlt+ is installed and set up according to the [installation guide](./installation.md)
- You are familiar with the [core concepts of dlt](../../reference/explainers/how-dlt-works.md)
- You have completed the [basic project tutorial](./tutorial.md)

Additionally, install the required Python packages:

```sh
pip install pandas numpy pyarrow streamlit dlt[duckdb] uv

```

## Implicit entities

By default, dlt+ will automatically create entities such as datasets or destinations when they are requested by the user or the executed code..
To see how this works, initialize a new project with: 
```sh

> dlt project init arrow duckdb --package my_dlt_project

Reviewer: if we do, let's explain why

<Link/>

In this tutorial, you will explore advanced features of a dlt+ project. By the end, you will be able to:
* Package a dlt+ project for distribution to other teams or stakeholders within your organization

## Prerequisites

Before you begin, ensure the following requirements are met:

- dlt+ is installed and set up according to the [installation guide](./installation.md)
- You are familiar with the [core concepts of dlt](../../reference/explainers/how-dlt-works.md)
- You have completed the [basic project tutorial](./tutorial.md)

Additionally, install the required Python packages:

```sh
pip install pandas numpy pyarrow streamlit dlt[duckdb] uv
```

## Packaging a project

`dlt+` provides tools to help you package a project for distribution.  
This makes your project installable via `pip` and easier to share across your organization.

> To create a packaged project, add the `--package` option when initializing:

Reviewer 2: ```suggestion
To create the project structure required for a package, add the `--package` option when initializing:
```

```sh
dlt project init arrow duckdb --package my_dlt_project
```

This creates the same basic project as in the tutorial, but places it inside a module named `my_dlt_project`, and includes a basic `pyproject.toml` file following PEP standards. 
You’ll also get a default `__init__.py` file to make the package usable after installation:

```sh
.
├── my_dlt_project/       # Your project module
│   ├── __init__.py       # Package entry point
│   ├── dlt.yml           # dlt+ project manifest
│   └── ...               # Other project files
├── .gitignore
└── pyproject.toml        # the main project manifest
```

Your `dlt.yml` works exactly the same as in non-packaged projects. 
The key difference is the module structure and the presence of the `pyproject.toml` file. 
The file includes a special entry point setting to let dlt+ discover your project:

```toml
[project.entry-points.dlt_package]
dlt-project = "my_project"
```

You can still run the pipeline as usual with the cli commands from the root folder:

```sh
dlt pipeline my_pipeline run
```

If you open the `__init__.py` file inside your project module, you'll see the full interface that users of your package will interact with.

> This interface is very similar to the `current` interface used in flat (non-packaged) projects. The main difference is that it automatically uses the `access` profile by default.

Reviewer 2: Can you please link to the documentation on the "current" interface - let me know if you need help figuring out how to link there

You can customize the `__init__.py` file to your project's needs.

### Using the packaged project

To demonstrate how your packaged project can be used, let's simulate a real-world scenario where a data scientist installs and runs your project in a separate Python environment.
In this example, we'll use the **uv** package manager, but the same steps apply when using **poetry** or **pip**.

> Assume your packaged `dlt+` project is located at: `/Volumes/my_drive/my_folder/pyproject.toml`. 

Reviewer 2: ```suggestion
Assume your packaged dlt+ project is located at: `/Volumes/my_drive/my_folder/pyproject.toml`. 
```

Navigate to a new directory and initialize your project:

```sh
uv init
```

Install your packaged project directly from the local path:

```sh
uv pip install /Volumes/my_drive/my_folder
```

Your dlt+ project is now available for use in this environment.

> Now, create a new Python file named `test_project.py` and use your packaged project like this:

Reviewer 2: ```suggestion
As an example, create a new Python file named `test_project.py` and use your packaged project like this:
```
